### ISSUE TITLE
[FEATURE] Standardize `state` field in billing and shipping address

### ISSUE THREAD

**Issue by sai-harsha-vardhan** [juspay/hyperswitch#2939]
### Feature Description

Currently, we accept the `state` field in both shipping and billing addresses as a string. However, in a few connectors like Cybersource, we have a mapping from state names to state codes. We propose standardizing the state field to accept only ISO 3166-2 codes (https://en.wikipedia.org/wiki/ISO_3166-2)

### Possible Implementation

We can use some crate that helps in validating ISO 3166-2 codes and accept these codes in payment request

### Have you spent some time checking if this feature request has been raised before?

- [X] I checked and didn't find a similar issue

### Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### Are you willing to submit a PR?

None

**Comment by Harshil-Jani** [comment]
@sai-harsha-vardhan  Using https://crates.io/crates/codes-iso-3166/0.1.3 would just make sense right ?

**Comment by sai-harsha-vardhan** [comment]
> @sai-harsha-vardhan Using https://crates.io/crates/codes-iso-3166/0.1.3 would just make sense right ?

Yes, I believe this should work. We need to ensure that the crate supports state codes as well, in addition to country codes

**Comment by Harshil-Jani** [comment]
> We need to ensure that the crate supports state codes as well, in addition to country codes

Awesome. I think the crate [rust_iso3166](https://crates.io/crates/rust_iso3166) is perfect in the case and also it is comparatively well-maintained. I have tried doing a demo run as below

```rust
use rust_iso3166::iso3166_2::from_code;
fn main() {
    let country_code = "IN-DL";
    let subdivision = from_code(country_code).unwrap();
    println!("{}: {:#?}", country_code, subdivision);
}
```
Output :
```rust
IN-DL: Subdivision {
    name: "Delhi",
    subdivision_type: "Union territory",
    code: "IN-DL",
    country_name: "India",
    country_code: "IN",
    region_code: "IN-DL",
}

```

We can just use `subdivision.name` to map the code with name.
If this is desired behaviour, Can you please elaborate a bit on the state field ? I cannot find it on regular searching in codebase related to billing or shipping address.

**Comment by sai-harsha-vardhan** [comment]
Hey @Harshil-Jani, this is awesome! You can find `state` in PaymentsRequest > billing / shipping > AddressDetails > state in api_models/payments. Do you wanna take this up?

**Comment by Harshil-Jani** [comment]
@sai-harsha-vardhan Sure, I would give it a try. I have few questions tho which were raise while I was trying to play with the code.

- Is the state field encrypted for a reason ? I see that country is not encrypted and thus If state should not be a Secret string then I was thinking of a new struct `StateCode` and which would hold the actual code string on which we can implement `validate` method which can be used further at all places where state is assigned. 
- Country comes from an internal enum. So, Shall we also deprecate the enum and use the crate for balancing things out for country and state/province ? Or other way round, State needs to go in enums and use no crates ?

**Comment by sai-harsha-vardhan** [comment]
Hey @Harshil-Jani, Currently, we are focusing solely on the state field and do not want to modify the country code, as it is being used for several other functionalities. We can address it as a separate issue. However, for the state, since it is Personally Identifiable Information (PII), we want to keep it secret. As there are a lot of state codes to maintain, it's better to use some crate and validate the request. Let me know your views on this. Thank you!
